File: p-isa_tools/data_formats/src/data_formats/python/heracles/data/io.py
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0


# TODO: create also C++ variants of below; given how simple and stable these functions should be just in replicated form, not shared code

import heracles.proto.data_pb2 as hpd
import heracles.proto.common_pb2 as hpc
import heracles.proto.fhe_trace_pb2 as hpf
import google.protobuf.json_format as gpj
import json
import heracles.data.transform as hdt
from google.protobuf.json_format import MessageToJson, MessageToDict
from glob import glob
import re
import sys

# load & store functions
# ===============================


def parse_manifest(filename: str) -> dict:
    manifest = dict()
    with open(filename, "r") as fp:
        cur_field = None
        found_first_field = False
        for linenum, cur_line in enumerate(fp):
            cur_line = cur_line.rstrip()
            if cur_line.startswith("[") and cur_line.endswith("]"):
                cur_field = cur_line[1:-1]
                found_first_field = True
                manifest[cur_field] = dict()
                continue

            if not found_first_field:
                continue
            cur_line_split = cur_line.split("=")
            if len(cur_line_split) != 2:
                print(
                    f"Warning: ignoring incorrect format in line {linenum}",
                    file=sys.stderr,
                )
                continue
            manifest[cur_field][cur_line_split[0]] = cur_line_split[1]

        if not found_first_field:
            raise Exception(f"Incorrect manifest file format: {filename}")

    return manifest


def generate_manifest(filename: str, manifest: dict):
    with open(filename, "w") as fp:
        for field, values in manifest.items():
            fp.write(f"[{field}]\n")
            for key, fn in values.items():
                fp.write(f"{key}={fn}\n")


# re-check
def store_hec_context_json(filename: str, context: hpd.FHEContext):
    print(
        f"Warning: Dumping FHE Context data trace to json can take a long time",
        file=sys.stderr,
    )
    with open(filename, "w") as fp:
        json.dump(MessageToDict(context), fp)


# re-check
def store_testvector_json(filename: str, testvector: hpd.TestVector):
    print(
        f"Warning: Dumping TestVector data trace to json can take a long time",
        file=sys.stderr,
    )
    with open(filename, "w") as fp:
        json.dump(MessageToDict(testvector), fp)


def load_hec_context_from_manifest(manifest: dict) -> hpd.FHEContext:
    context_base_fn = manifest["context"]["main"]
    context_pb = hpd.FHEContext()

    with open(context_base_fn, "rb") as fp:
        context_pb.ParseFromString(fp.read())

    if "rotation_keys" in manifest:
        for ge, gk_fn in manifest["rotation_keys"].items():
            gk_pb = hpd.KeySwitch()
            with open(gk_fn, "rb") as fp:
                gk_pb.ParseFromString(fp.read())
            context_pb.ckks_info.keys.rotation_keys[int(ge)].CopyFrom(gk_pb)

    return context_pb


def store_hec_context(filename: str, context_pb: hpd.FHEContext) -> dict:
    hec_context_manifest = {"context": dict()}
    tmp_context = hpd.FHEContext()
    tmp_context.CopyFrom(context_pb)

    if tmp_context.ByteSize() > 1 << 30:
        hec_context_manifest["rotation_keys"] = dict()
        for gkct, (ge, gk_pb) in enumerate(
            tmp_context.ckks_info.keys.rotation_keys.items()
        ):
            parts_fn = f"{filename}_hec_context_part_{gkct + 1}"
            hec_context_manifest["rotation_keys"][ge] = parts_fn
            with open(parts_fn, "wb") as fp:
                fp.write(gk_pb.SerializeToString())
        tmp_context.ckks_info.keys.ClearField("rotation_keys")

    main_fn = f"{filename}_hec_context_part_0"
    hec_context_manifest["context"]["main"] = main_fn
    with open(main_fn, "wb") as fp:
        fp.write(tmp_context.SerializeToString())

    return hec_context_manifest


def load_testvector_from_manifest(manifest: dict) -> hpd.TestVector:
    # segmented
    testvector_pb = hpd.TestVector()
    if len(manifest["testvector"]) > 1:
        for sym, parts_fn in manifest["testvector"].items():
            data = hpd.Data()
            with open(parts_fn, "rb") as fp:
                data.ParseFromString(fp.read())
            testvector_pb.sym_data_map[sym].CopyFrom(data)
    # whole
    else:
        full_fn = manifest["testvector"]["full"]
        with open(full_fn, "rb") as fp:
            testvector_pb.ParseFromString(fp.read())

    return testvector_pb


def store_testvector(filename: str, testvector_pb: hpd.TestVector) -> dict:
    testvector_manifest = {"testvector": dict()}
    if testvector_pb.ByteSize() > 1 << 30:
        for tvct, (sym, data_pb) in enumerate(testvector_pb.sym_data_map.items()):
            parts_fn = f"{filename}_testvector_part_{tvct}"
            testvector_manifest["testvector"][sym] = parts_fn
            with open(parts_fn, "wb") as fp:
                fp.write(data_pb.SerializeToString())
    else:
        full_fn = f"{filename}_testvector_part_0"
        testvector_manifest["testvector"]["full"] = full_fn
        with open(full_fn, "wb") as fp:
            fp.write(testvector_pb.SerializeToString())

    return testvector_manifest


def load_hec_context(filename: str) -> hpd.FHEContext:
    manifest = parse_manifest(filename)
    return load_hec_context_from_manifest(manifest)


def load_testvector(filename: str) -> hpd.FHEContext:
    manifest = parse_manifest(filename)
    return load_testvector_from_manifest(manifest)


def load_data_trace(filename: str) -> tuple[hpd.FHEContext, hpd.TestVector]:
    manifest = parse_manifest(filename)
    return (
        load_hec_context_from_manifest(manifest),
        load_testvector_from_manifest(manifest),
    )


def store_data_trace(
    filename: str, context_pb: hpd.FHEContext, testvector_pb: hpd.TestVector
):
    generate_manifest(
        filename,
        {
            **store_hec_context(filename, context_pb),
            **store_testvector(filename, testvector_pb),
        },
    )

File: p-isa_tools/data_formats/cpp/heracles/data/io.cpp
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "heracles/data/io.h"
#include <algorithm>
#include <cctype>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <google/protobuf/util/json_util.h>
#include <iostream>

namespace heracles::data
{
hdf_manifest parse_manifest(const std::string &filename)
{
    hdf_manifest manifest;
    std::ifstream file(filename);

    if (!file.is_open())
    {
        throw std::runtime_error("Manifest file not found: " + filename);
    }

    std::string header, current_line;
    auto trim = [](const std::string &line) {
        std::string res = line;
        res.erase(std::remove_if(res.begin(), res.end(), isspace), res.end());
        return res;
    };

    std::string cur_field = "";
    bool found_first_field = false;
    int linenum = 0;
    while (std::getline(file, current_line))
    {
        ++linenum;
        current_line = trim(current_line);
        if (current_line.front() == '[' && current_line.back() == ']')
        {
            cur_field = current_line.substr(1, current_line.size() - 2);
            found_first_field = true;
            continue;
        }

        if (!found_first_field)
        {
            continue;
        }

        std::stringstream current_line_ss(current_line);
        std::string value;
        std::vector<std::string> values;
        while (std::getline(current_line_ss, value, '='))
        {
            values.push_back(trim(value));
        }
        // if format is not "a=b", pass
        if (values.size() != 2)
        {
            std::cout << "Warning : ignoring incorrect format in line :" << linenum << std::endl;
            continue;
        }

        manifest[cur_field][values[0]] = values[1];
    }

    if (!found_first_field)
        throw std::runtime_error("Incorrect manifest format: " + filename);

    return manifest;
}

void generate_manifest(const std::string &filename, const hdf_manifest &manifest)
{
    std::stringstream ss;
    for (const auto &[field, values] : manifest)
    {
        ss << "[" << field << "]" << std::endl;
        for (const auto &[key, fn] : values)
        {
            ss << key << "=" << fn << std::endl;
        }
    }
    std::ofstream ofile(filename, std::ios::out);
    ofile << ss.rdbuf();
}

bool store_hec_context_json(const std::string &filename, const heracles::data::FHEContext &context)
{
    std::string json_str;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    options.always_print_primitive_fields = true;
    auto rc = ::google::protobuf::util::MessageToJsonString(context, &json_str, options);
    std::ofstream json_ofile(filename, std::ios::out);
    json_ofile << json_str;

    return true;
}

bool store_testvector_json(const std::string &filename, const heracles::data::TestVector &test_vector)
{
    std::string json_str;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    options.always_print_primitive_fields = true;
    auto rc = ::google::protobuf::util::MessageToJsonString(test_vector, &json_str, options);
    std::ofstream json_ofile(filename, std::ios::out);

    json_ofile << json_str;
    return true;
}

void store_hec_context(
    hdf_manifest *manifest_out, const std::string &filename, const heracles::data::FHEContext &context_pb)
{
    auto tmp_context = context_pb;
    if (context_pb.ByteSizeLong() > (1 << 30))
    {
        int gkct = 1;
        for (const auto &[ge, key] : tmp_context.ckks_info().keys().rotation_keys())
        {
            std::stringstream parts_fnss;
            parts_fnss << filename << "_hec_context_part_" << gkct++;
            std::ofstream pb_ofile(parts_fnss.str(), std::ios::out | std::ios::binary);
            (*manifest_out)["rotation_keys"][std::to_string(ge)] = parts_fnss.str();
            if (!key.SerializeToOstream(&pb_ofile))
                throw std::runtime_error("Serializing rotation key failed");
        }
        tmp_context.mutable_ckks_info()->mutable_keys()->clear_rotation_keys();
    }

    auto main_fs = filename + "_hec_context_part_0";
    (*manifest_out)["context"]["main"] = main_fs;
    std::ofstream pb_ofile(main_fs, std::ios::out | std::ios::binary);
    if (!tmp_context.SerializeToOstream(&pb_ofile))
        throw std::runtime_error("Serializing main hec context failed");
}
void store_testvector(
    hdf_manifest *manifest_out, const std::string &filename, const heracles::data::TestVector &testvector_pb)
{
    if (testvector_pb.ByteSizeLong() > (1 << 30))
    {
        int tvct = 0;
        for (const auto &[sym, data_part] : testvector_pb.sym_data_map())
        {
            std::stringstream parts_fnss;
            parts_fnss << filename << "_testvector_part_" << tvct++;
            auto parts_fn = parts_fnss.str();
            (*manifest_out)["testvector"][sym] = parts_fn;
            std::ofstream pb_ofile(parts_fn, std::ios::out | std::ios::binary);
            if (!data_part.SerializeToOstream(&pb_ofile))
                throw std::runtime_error("Serializing test vector part " + sym + " failed. File : " + parts_fn);
        }
        return;
    }

    auto full_fn = filename + "_testvector_part_0";
    (*manifest_out)["testvector"]["full"] = full_fn;
    std::ofstream pb_ofile(full_fn, std::ios::out | std::ios::binary);
    if (!testvector_pb.SerializeToOstream(&pb_ofile))
        throw std::runtime_error("Serializing full test vector failed. File : " + full_fn);
}

bool store_data_trace(
    const std::string &filename, const heracles::data::FHEContext &context_pb,
    const heracles::data::TestVector &testvector_pb)
{
    hdf_manifest manifest_datatrace;
    try
    {
        store_hec_context(&manifest_datatrace, filename, context_pb);
        store_testvector(&manifest_datatrace, filename, testvector_pb);
        generate_manifest(filename, manifest_datatrace);
    }
    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during store_data_trace, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }

    return true;
}

void load_hec_context_from_manifest(heracles::data::FHEContext *context_pb, const hdf_manifest &manifest)
{
    try
    {
        std::filesystem::path main_fn(manifest.at("context").at("main"));
        std::ifstream context_pb_ifile(main_fn, std::ios::in | std::ios::binary);
        context_pb->ParseFromIstream(&context_pb_ifile);

        if (manifest.count("rotation_keys"))
        {
            for (const auto &[ge, gk_fn] : manifest.at("rotation_keys"))
            {
                heracles::data::KeySwitch gk_pb;
                std::ifstream gk_pb_ifile(gk_fn, std::ios::in | std::ios::binary);
                gk_pb.ParseFromIstream(&gk_pb_ifile);
                (*(context_pb->mutable_ckks_info()
                       ->mutable_keys()
                       ->mutable_rotation_keys()))[static_cast<std::uint32_t>(std::stoul(ge))] = gk_pb;
            }
        }
    }

    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during load_hec_context, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }
}
void load_testvector_from_manifest(heracles::data::TestVector *testvector_pb, const hdf_manifest &manifest)
{
    try
    {
        if (manifest.at("testvector").find("full") != manifest.at("testvector").end())
        { // single file
            const auto &full_fn = manifest.at("testvector").at("full");
            std::ifstream pb_ifile(full_fn, std::ios::in | std::ios::binary);
            testvector_pb->ParseFromIstream(&pb_ifile);
        }
        else
        { // segmented
            for (const auto &[sym, parts_fn] : manifest.at("testvector"))
            {
                std::ifstream pb_ifile(parts_fn, std::ios::in | std::ios::binary);
                (*testvector_pb->mutable_sym_data_map())[sym].ParseFromIstream(&pb_ifile);
            }
        }
    }
    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during _load_testvector, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }
}

heracles::data::FHEContext load_hec_context(const std::string &filename)
{
    heracles::data::FHEContext context_pb;
    try
    {
        auto manifest = parse_manifest(filename);
        load_hec_context_from_manifest(&context_pb, manifest);
    }
    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during load_data_trace, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }

    return context_pb;
}
heracles::data::TestVector load_testvector(const std::string &filename)
{
    heracles::data::TestVector testvector_pb;
    try
    {
        auto manifest = parse_manifest(filename);
        load_testvector_from_manifest(&testvector_pb, manifest);
    }
    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during load_data_trace, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }

    return testvector_pb;
}

std::pair<heracles::data::FHEContext, heracles::data::TestVector> load_data_trace(const std::string &filename)
{
    heracles::data::FHEContext context_pb;
    heracles::data::TestVector testvector_pb;
    try
    {
        auto manifest = parse_manifest(filename);
        load_hec_context_from_manifest(&context_pb, manifest);
        load_testvector_from_manifest(&testvector_pb, manifest);
    }
    catch (const std::runtime_error &err)
    {
        std::cerr << "Runtime error during load_data_trace, err: " << err.what() << std::endl;
        throw err;
    }
    catch (...)
    {
        std::cerr << "Unknown exception caught in " << __FUNCTION__ << "in file" << __FILE__ << std::endl;
        throw;
    }

    return { context_pb, testvector_pb };
}

} // namespace heracles::data
